pinot-core/src/main/java/org/apache/pinot/core/query/pruner/ColumnValueSegmentPruner.java

```diff
@@ -99,6 +102,11 @@ public List<IndexSegment> prune(List<IndexSegment> segments, QueryContext query)
 
     int numSegments = segments.size();
     List<IndexSegment> selectedSegments = new ArrayList<>(numSegments);
+
+    // By using an IdentityHashMap, we may be creating more entries, but each lookup will be faster
+    // which should compensate in cases where there are hundreds of segments
+    Map<Object, Hash128> hashCache = new IdentityHashMap<>();
```

Comment:
IdentityMap removes the call to equals, which can be relatively expensive when the keys (specially strings) are not the same reference but has the same hash code. This is can happen when an expression contains the same literal several times. For example, in the following query, the literal "someString" may be the same instance or a different one:

```sql
select * from Table where col1 = "someString" and col2 = "someString"
```

Reply:
Do you have some data to justify this? Computing identity hash codes have various side effects and it's not always a win.

Reply:
As shown by @Jackie-Jiang, in the equals case it is not correct to use the identity, so I'm going to use a normal map here.